src/reason-parser/reason_toolchain.ml

```diff
@@ -719,21 +735,11 @@ module Reason_syntax = struct
       (* Enter error recovery state *)
       handle_other supplier checkpoint
 
-    | I.HandlingError env ->
-      (* If not in a recoverable state, fail early by raising a
-       * customized Error object *)
-      let loc = last_token_loc supplier in
-      let token = match supplier.last_token with
-        | Some token -> token
-        | None -> assert false
-      in
-      let state = I.current_state_number env in
-      (* Check the error database to see what's the error message
-       * associated with the current parser state *)
-      let msg = Reason_parser_explain.message env token in
-      let msg_with_state = Printf.sprintf "%d: %s" state msg in
-      raise (Reason_syntax_util.Error (loc, (Reason_syntax_util.Syntax_error msg_with_state)))
-
+    | I.HandlingError _env as error_checkpoint ->
+      (* If not in a recoverable state, resume just enough to be able to
+       * catch a nice error message above. *)
+      let cp = I.resume error_checkpoint in
```

Comment:
Are there cases where we don't want to resume when we're handling an error?

Reply:
I think there are, but I'm not sure which heuristics to go for here.